pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,246 @@
+package nodelabeller
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController     *device_manager.DeviceController
+	configMapInformer cache.SharedIndexInformer
+	nodeInformer      cache.SharedIndexInformer
+	clientset         kubecli.KubevirtClient
+	Queue             workqueue.RateLimitingInterface
+	host              string
+	namespace         string
+	logger            *log.FilteredLogger
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, nodeInformer, configMapInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		configMapInformer: configMapInformer,
+		nodeInformer:      nodeInformer,
+		clientset:         clientset,
+		Queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:      host,
+		namespace: namespace,
+		logger:    log.DefaultLogger(),
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
```

Comment:
/hold

Let's talks about this line. It is explicitly allowed to run virt-handler on a node which does not  yet have kvm present. virt-handler right now reports the node as unschedulable for VMs, but as soon as kvm is there it will report it as schedulable. Will this behaviour be preserved?

Reply:
When kvm is not present, init container will exit with 0 (https://github.com/kubevirt/kubevirt/pull/3133/files#diff-2f33df9ddfbca886622a215ae805b6f3R6). Then labeller will exit, because it will not have any info about supported cpus. Issue is, that if node gets kvm, labeller cannot relabel the node, because all info is provided by virt-launcher in init container. The virt-handler would have to restart to run init container.